shell/platform/android/io/flutter/embedding/android/KeyboardManager.java

```diff
@@ -0,0 +1,204 @@
+ *       only this time the {@link KeyboardManager} will not try to handle the redispatched {@link
+ *       KeyEvent}.
+ * </ul>
+ */
+public class KeyboardManager {
+  private static final String TAG = "KeyboardManager";
+
+  KeyboardManager(
+      View view, @NonNull TextInputPlugin textInputPlugin, PrimaryResponder[] primaryResponders) {
+    this.view = view;
+    this.textInputPlugin = textInputPlugin;
+    this.primaryResponders = primaryResponders;
+  }
+
+  /**
+   * Constructor for {@link KeyboardManager} that uses a {@link KeyChannelResponder} as its only
+   * {@link PrimaryResponder}.
+   *
+   * <p>The view is used as the destination to send the synthesized key to. This means that the the
+   * next thing in the focus chain will get the event when the {@link KeyChannelResponder} returns
+   * false from onKeyDown/onKeyUp.
+   *
+   * <p>It is possible that that in the middle of the async round trip, the focus chain could
+   * change, and instead of the native widget that was "next" when the event was fired getting the
+   * event, it may be the next widget when the event is synthesized that gets it. In practice, this
+   * shouldn't be a huge problem, as this is an unlikely occurrence to happen without user input,
+   * and it may actually be desired behavior, but it is possible.
+   *
+   * @param view takes the activity to use for re-dispatching of events that were not handled by the
+   *     framework.
+   * @param textInputPlugin a plugin, which, if set, is given key events before the framework is,
+   *     and if it has a valid input connection and is accepting text, then it will handle the event
+   *     and the framework will not receive it.
+   * @param keyEventChannel the event channel to listen to for new key events.
+   */
+  public KeyboardManager(
+      View view, @NonNull TextInputPlugin textInputPlugin, KeyEventChannel keyEventChannel) {
+    this(
+        view,
+        textInputPlugin,
+        new KeyboardManager.PrimaryResponder[] {new KeyChannelResponder(keyEventChannel)});
+  }
+
+  /**
+   * The interface for responding to a {@link KeyEvent} asynchronously.
+   *
+   * <p>Implementers of this interface should be owned by a {@link KeyboardManager}, in order to
+   * receive key events.
+   *
+   * <p>After receiving a {@link KeyEvent}, the {@link PrimaryResponder} must call the supplied
+   * {@link OnKeyEventHandledCallback} exactly once, to inform the {@link KeyboardManager} whether
+   * it wishes to handle the {@link KeyEvent}. The {@link KeyEvent} will not be propagated to the
+   * {@link TextInputPlugin} or be redispatched to the view hierachy if the key responder answered
+   * yes.
+   *
+   * <p>If a {@link PrimaryResponder} fails to call the {@link OnKeyEventHandledCallback} callback,
+   * the {@link KeyEvent} will never be sent to the {@link TextInputPlugin}, and the {@link
+   * KeyboardManager} class can't detect such errors as there is no timeout.
```

Comment:
This is mostly for debugging. Knowing the symptoms would probably help people with debugging, since there's no good way to catch the programmer error in our implementation.